packages/components/src/internal/components/chart/ChartBuilderModal.tsx

```diff
@@ -0,0 +1,648 @@
+
+import { naturalSortByProperty } from '../../../public/sort';
+
+import { LoadingSpinner } from '../base/LoadingSpinner';
+
+import { deleteChart, saveChart } from '../../actions';
+
+import { QueryModel } from '../../../public/QueryModel/QueryModel';
+import { RequiresModelAndActions } from '../../../public/QueryModel/withQueryModels';
+
+import { useServerContext } from '../base/ServerContext';
+import { hasPermissions } from '../base/models/User';
+
+import { Alert } from '../base/Alert';
+import { FormButtons } from '../../FormButtons';
+
+import { getContainerFilter } from '../../query/api';
+
+import { SVGIcon } from '../base/SVGIcon';
+
+import { ChartConfig, ChartQueryConfig, GenericChartModel } from './models';
+
+interface AggregateFieldInfo {
+    name: string;
+    value: string;
+}
+
+export interface ChartFieldInfo {
+    aggregate?: AggregateFieldInfo;
+    // allowMultiple?: boolean;
+    label: string;
+    name: string;
+    nonNumericOnly?: boolean;
+    numericOnly?: boolean;
+    required: boolean;
+}
+
+export interface ChartTypeInfo {
+    fields: ChartFieldInfo[];
+    hidden?: boolean;
+    imgUrl: string;
+    name: string;
+    title: string;
+}
+
+const HIDDEN_CHART_TYPES = ['time_chart'];
+const RIGHT_COL_FIELDS = ['color', 'shape', 'series'];
+export const MAX_ROWS_PREVIEW = 100000;
+export const MAX_POINT_DISPLAY = 10000;
+const BLUE_HEX_COLOR = '3366FF';
+
+const ICONS = {
+    bar_chart: 'bar_chart',
+    box_plot: 'box_plot',
+    pie_chart: 'pie_chart',
+    scatter_plot: 'xy_scatter',
+    line_plot: 'xy_line',
+};
+
+interface Props extends RequiresModelAndActions {
+    onHide: (successMsg?: string) => void;
+    savedChartModel?: GenericChartModel;
+}
+
+export const ChartBuilderModal: FC<Props> = memo(({ actions, model, onHide, savedChartModel }) => {
+    const CHART_TYPES = LABKEY_VIS?.GenericChartHelper.getRenderTypes();
+    const divId = useMemo(() => generateId('chart-'), []);
+    const ref = useRef<HTMLDivElement>(undefined);
+    const { user } = useServerContext();
+    const canShare = useMemo(
+        () => savedChartModel?.canShare ?? hasPermissions(user, [PermissionTypes.ShareReportPermission]),
+        [savedChartModel, user]
+    );
+    const chartTypes: ChartTypeInfo[] = useMemo(
+        () => CHART_TYPES.filter(type => !type.hidden && !HIDDEN_CHART_TYPES.includes(type.name)),
+        [CHART_TYPES]
+    );
+    const containerFilter = useMemo(() => getContainerFilter(model.containerPath), [model.containerPath]);
+
+    const [loadingData, setLoadingData] = useState<boolean>(false);
+    const [saving, setSaving] = useState<boolean>(false);
+    const [deleting, setDeleting] = useState<boolean>(false);
+    const [showConfirmDelete, setShowConfirmDelete] = useState<boolean>(false);
+    const [error, setError] = useState<string>();
+    const [previewMsg, setPreviewMsg] = useState<string>();
+    const [reportConfig, setReportConfig] = useState<Record<string, any>>();
```

Comment:
Is `Record<string, any>` appropriate here? Is the report config so dynamic we have to use `any`?

Reply:
I changed it to define `SaveReportConfig`